### ISSUE TITLE
[YSQL] Unexpected `Restart read required` error

### ISSUE THREAD

**Issue by d-uspenskiy** [yugabyte/yugabyte-db#13387]
Jira Link: [DB-3021](https://yugabyte.atlassian.net/browse/DB-3021)
### Description

The unit test below fails on `release` build (`./yb_build.sh release --gtest_filter PgMiniTest.MegaTest -n 20`) with the following error

```
2022-07-21 10:18:11.737 UTC [988759] ERROR:  Query error: Restart read required at: { read: { days: 19194 time: 10:18:11.736547 } local_limit: { days: 19194 time: 10:18:11.736547 } global_limit: <min> in_txn_limit: <max> serial_no: 0 }
2022-07-21 10:18:11.737 UTC [988759] STATEMENT:  alter table defpart_attach_test_d add check (a > 1)
../../src/yb/yql/pgwrapper/pg_mini-test.cc:2847: Failure
Failed
Bad status: Network error (yb/yql/pgwrapper/libpq_utils.cc:266): Execute of 'alter table defpart_attach_test_d add check (a > 1)' failed: 7, message: ERROR:  Query error: Restart read required at: { read: { days: 19194 time: 10:18:11.736547 } local_limit: { days: 19194 time: 10:18:11.736547 } global_limit: <min> in_txn_limit: <max> serial_no: 0 } (pgsql error 40001) (aux msg ERROR:  Query error: Restart read required at: { read: { days: 19194 time: 10:18:11.736547 } local_limit: { days: 19194 time: 10:18:11.736547 } global_limit: <min> in_txn_limit: <max> serial_no: 0 })
```

```
TEST_F(PgMiniTest, YB_DISABLE_TEST_IN_TSAN(MegaTest)) {
  auto conn = ASSERT_RESULT(Connect());
  for (int i = 0; i < 20; ++i) {
    ASSERT_OK(conn.Execute("create table defpart_attach_test (a int) partition by list (a)"));
    ASSERT_OK(conn.Execute("create table defpart_attach_test1 partition of defpart_attach_test for values in (1)"));
    ASSERT_OK(conn.Execute("create table defpart_attach_test_d (a int)"));
    ASSERT_OK(conn.Execute("insert into defpart_attach_test_d values (1), (2)"));

    ASSERT_NOK(conn.Execute("alter table defpart_attach_test attach partition defpart_attach_test_d default"));
    ASSERT_OK(conn.Execute("delete from defpart_attach_test_d where a = 1"));
    ASSERT_OK(conn.Execute("alter table defpart_attach_test_d add check (a > 1)")); // <-- this statement fails

    ASSERT_OK(conn.Execute("drop table defpart_attach_test_d, defpart_attach_test1, defpart_attach_test"));
  }
}
```

Stack trace for the error is the following

```
I0721 10:34:42.541097 1041100 pg_yb_utils.c:453] HandleYBStatus: Query error: Restart read required at: { read: { days: 19194 time: 10:34:42.540834 } local_limit: { days: 19194 time: 10:34:42.540834 } global_limit: <min> in_txn_limit: <max> serial_no:
 0 }
    @           0x9de2d0  HandleYBStatusAtErrorLevel
    @           0x56c547  ybc_getnext_heaptuple
    @           0x56cda1  ybc_heap_getnext
    @           0x680552  ATRewriteTable
    @           0x66868d  ATController
    @           0x85752f  ProcessUtilitySlow
    @           0x8562a0  standard_ProcessUtility
    @           0x85588e  YBProcessUtilityDefaultHook
    @     0x7ff345201e40  pgss_ProcessUtility
    @     0x7ff3451f8aa7  ybpgm_ProcessUtility
    @     0x7ff3451dc6fe  pg_hint_plan_ProcessUtility
    @           0x9e1cd7  YBTxnDdlProcessUtility
    @           0x8554da  PortalRunUtility
    @           0x854bc7  PortalRunMulti
    @           0x854486  PortalRun
    @           0x851c06  yb_exec_simple_query_impl
    @           0x8521a6  yb_exec_query_wrapper_one_attempt
    @           0x84ed1b  PostgresMain
    @           0x7bae1c  BackendRun
    @           0x7ba270  ServerLoop
    @           0x7b682b  PostmasterMain
    @           0x717a3d  PostgresServerProcessMain
    @           0x717ee2  main
    @     0x7ff3499ff825  __libc_start_main
    @           0x4c63d9  _start
```

Possible reason of the problem might be the reading from users table in context of DDL transaction in case this table was modified by previous DML transaction statement in context of same connection. These DML and DDL transactions might use different status tablet servers which has skewed clock.

**Comment by averyshum62** [comment]
The regress test file yb_pg_alter_table.sql sometimes fails with this error, `Query error: Restart read required`. The failing portion of the test is currently not included, but should be imported when this issue is resolved.

**Comment by siddharth2411** [comment]
Facing same error while running the test file: _yb_pg_stat.sql_ Particularly it is failing while executing the CREATE TABLE statement in the below code:

`create database colocated_db with colocated = true;
\c colocated_db;
create table mycolocatedtable (c1 INT PRIMARY KEY, c2 TEXT, c3 INT);
insert into mycolocatedtable (c1, c2, c3) values (6, '9', 8);
create index mycolocatedtable_index1 on mycolocatedtable (c2);
create index mycolocatedtable_index2 on mycolocatedtable (c3);
select * from mycolocatedtable where c2='9';
select pg_sleep(1);
select indexrelname,idx_scan from pg_stat_user_indexes where indexrelname ~ 'mycolocatedtable_index.' order by (indexrelname);
select * from mycolocatedtable where c3=8;
select pg_sleep(1);
select indexrelname,idx_scan from pg_stat_user_indexes where indexrelname ~ 'mycolocatedtable_index.' order by (indexrelname);`

**Comment by jasonyb** [comment]
Also hit it on internal log https://jenkins.dev.yugabyte.com/job/yugabyte-db-phabricator/177294/artifact/java/yb-pgsql/target/surefire-reports_org.yb.pgsql.TestPgRegressAggregates__testPgRegressAggregates/org.yb.pgsql.TestPgRegressAggregates-output.txt.gz

```
*** ${TEST_TMPDIR}/pgregress_output/yb_pg_aggregates_schedule/expected/yb_aggregates.out  2023-08-09 03:25:45.114176859 +0000
--- ${TEST_TMPDIR}/pgregress_output/yb_pg_aggregates_schedule/results/yb_aggregates.out 2023-08-09 03:25:45.065176707 +0000
***************
*** 1979,2022 ****
  CREATE DATABASE co COLOCATION TRUE;
  \c co
  CREATE TABLE t (i int, j int, k int);
  CREATE INDEX NONCONCURRENTLY i ON t (j, k DESC, i);
...
--- 1979,2010 ----
  CREATE DATABASE co COLOCATION TRUE;
  \c co
  CREATE TABLE t (i int, j int, k int);
+ ERROR:  Restart read required at: { read: { days: 19578 time: 03:25:44.954729 } local_limit: { days: 19578 time: 03:25:44.954729 } global_limit: <min> in_txn_limit: <max> serial_no: 0 }
  CREATE INDEX NONCONCURRENTLY i ON t (j, k DESC, i);
+ ERROR:  relation "t" does not exist
...
```

**Comment by d-uspenskiy** [comment]
Unit test from description doesn't fail on latest master debug/release build. It is supposed that problem is solved in context of #14519
